File: app/src/main/java/com/example/uos_lms/core/domain/model/TeacherAttendanceRecord.java
package com.example.uos_lms.core.domain.model;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

/** Course-wise teacher attendance: was this subject's assigned teacher present for THIS course
 * on a given day - a teacher can be PRESENT for one course and ABSENT for another on the same
 * date, since each record is scoped to one (subject, dateKey) session. Distinct from
 * AttendanceRecord, which is a student's attendance in a subject's class session. */
@Getter
@Builder(toBuilder = true)
@EqualsAndHashCode
@ToString
public class TeacherAttendanceRecord {
    private final String id;
    private final String departmentId;
    private final String semesterId;
    private final String subjectId;
    private final String subjectCode;
    private final String subjectTitle;
    private final String dateKey;
    private final long dateMillis;
    private final String teacherUid;
    private final String teacherName;
    private final AttendanceStatus status;
    private final String markedBy;
    @Builder.Default
    private final long createdAt = 0L;
}

File: src/models/TeacherAttendanceRecord.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

// Course-wise teacher attendance: was this subject's assigned teacher present for THIS course
// on a given day - a teacher can be PRESENT for one course and ABSENT for another on the same
// date, since each record is scoped to one (subject, dateKey) session, not the whole day.
const teacherAttendanceRecordSchema = new Schema(
  {
    teacherId: { type: String, ref: 'User', required: true },
    subjectId: { type: Schema.Types.ObjectId, ref: 'Subject', required: true },
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', required: true },
    semesterId: { type: Schema.Types.ObjectId, ref: 'Semester', required: true },
    dateKey: { type: String, required: true },
    status: { type: String, enum: ['PRESENT', 'ABSENT'], required: true },
    markedBy: { type: String, ref: 'User', required: true },
  },
  { timestamps: true }
);

// One record per course per day - upserted, mirrors AttendanceRecord's idempotency pattern.
// Scoping the unique key to (subjectId, dateKey) rather than (teacherId, dateKey) is what lets
// the same teacher have independent PRESENT/ABSENT statuses across different courses same day.
teacherAttendanceRecordSchema.index({ subjectId: 1, dateKey: 1 }, { unique: true });

module.exports = mongoose.model('TeacherAttendanceRecord', teacherAttendanceRecordSchema);
